api/classes/class-lightweight-api.php

```diff
@@ -0,0 +1,303 @@
+			filter_input( INPUT_POST | INPUT_GET, 'suppress_all_newsletter_campaigns_if_one_dismissed', FILTER_SANITIZE_SPECIAL_CHARS ) :
+			null; // phpcs:ignore
+		$this->has_newsletter_prompt = ! empty( $_REQUEST['has_newsletter_prompt'] ) ? // phpcs:ignore
+			filter_input( INPUT_POST | INPUT_GET, 'has_newsletter_prompt', FILTER_SANITIZE_SPECIAL_CHARS ) :
+			null; // phpcs:ignore
+
+		$this->referer_url = filter_input( INPUT_SERVER, 'HTTP_REFERER', FILTER_SANITIZE_STRING );
+
+		if ( 'inline' !== $this->placement && 'always' === $this->frequency ) {
+			$this->frequency = 'once';
+		}
+	}
+
+	/**
+	 * Verify referer is valid.
+	 */
+	public function verify_referer() {
+		$http_referer = ! empty( $_SERVER['HTTP_REFERER'] ) ? parse_url( $_SERVER['HTTP_REFERER'] , PHP_URL_HOST ) : null; // phpcs:ignore
+		$valid_referers = [
+			$http_referer,
+			// TODO: Add AMP Cache.
+		];
+		$http_host = ! empty( $_SERVER['HTTP_HOST'] ) ? $_SERVER['HTTP_HOST'] : null; // phpcs:ignore
+		return in_array( strtolower( $http_host ), $valid_referers, true );
+	}
+
+	/**
+	 * Establish DB connection.
+	 */
+	public function connect() {
+		$credentials = $this->get_credentials();
+		if ( ! $credentials ) {
+			$this->error( 'no_credentials' );
+		}
+		$db = mysqli_init(); //phpcs:ignore
+		$db->real_connect(
+			$credentials['db_host'],
+			$credentials['db_user'],
+			$credentials['db_password'],
+			$credentials['db_name']
+		);
+		return $db;
+	}
+
+	/**
+	 * Get database credentials from environment variables (Atomic) or special config file.
+	 */
+	public function get_credentials() {
+		if ( getenv( 'DB_NAME' ) && getenv( 'DB_USER' ) && getenv( 'DB_PASSWORD' ) ) {
+			return [
+				'db_host'     => 'localhost',
+				'db_name'     => getenv( 'DB_NAME' ),
+				'db_user'     => getenv( 'DB_USER' ),
+				'db_password' => getenv( 'DB_PASSWORD' ),
+			];
+		}
+		$config = $_SERVER['DOCUMENT_ROOT'] . '/newspack-popups-config.php'; //phpcs:ignore
+		if ( file_exists( $config ) ) {
+			require $config;
+			return [
+				'db_host'     => DB_HOST,
+				'db_name'     => DB_NAME,
+				'db_user'     => DB_USER,
+				'db_password' => DB_PASSWORD,
+			];
+		}
+		return null;
+	}
+
+	/**
+	 * Get transient name.
+	 */
+	public function transient_name() {
+		return sprintf( '_transient_%s-%s-popup', $this->rid, $this->popup_id );
+	}
+
+	/**
+	 * Get suppression-related transient value
+	 *
+	 * @param string $prefix transient prefix.
+	 */
+	public function get_suppression_data_transient_name( $prefix ) {
+		return sprintf(
+			'_transient_%s-%s',
+			$prefix,
+			$this->rid
+		);
+	}
+
+	/**
+	 * Get data from transient.
+	 *
+	 * @param string $transient_name The transient's name.
+	 */
+	public function get_transient( $transient_name ) {
+		$query                    = sprintf(
+			"SELECT option_value FROM wp_options WHERE option_name = '%s'",
```

Comment:
I assume won't be necessary in the future when this has a dedicated table but just noting: we should handle different DB prefixes. It's not always going to be `wp_`. Maybe another constant in `newspack-popups-config.php`?

Reply:
If the plugin does the approach Ivan recommends where we parse the wp-config file, this could get the prefix from there.

Reply:
I know this is supposed to be bypassing WP, but the DB prefix is `$wpdb->prefix` – it could be read once and stored for later perhaps?

Reply:
I took a shot at extracting the prefix using a regex in https://github.com/Automattic/newspack-popups/pull/208/commits/cc16d6d0eaa5e015711ba4ea2cb2624010e25525. I'm becoming more and more concerned about extracting the DB credentials from `wp-config.php` by string manipulation though. To give one example of the potential brittleness, if there are commented-out constants the script gets confused, e.g.:

```
/*
define( 'DB_NAME', 'notreal' );
*/
define( 'DB_NAME', 'wordpresssimple' );
```

Reply:
> it could be read once and stored for later perhaps?

@adekbadek how would this work?

Reply:
The plugin could generate a file – unless there's already one – with some initial WP-related data. That could be the only point of contact with WP.

Reply:
Take a look at this [draft PR](https://github.com/Automattic/newspack-popups/pull/210), which changes back to using  a custom config file. We could look at auto-generating this file when the plugin is activated.

Reply:
Config file solves this problem nicely.